### PR TITLE
Fix recent error and refactor role for clarity and greater flexibility

### PR DESCRIPTION
**Pull request by allthatilk**
This will need testing out as I put this together pretty quickly after work and I don't have time to test it, but you should just be able to spin up a DO droplet and check the authorised key file after running it.

So I noticed that the most recent commit seems to misunderstand how the keys are being added to the authorised keys file. It's not pulling individual items from the files, it's overwriting the whole file. Running this role in its current state would mean that only the keys for the devs in the external file will be added, and if you're an internal IATI dev running the playbook, the playbook would cancel out as soon as the external dev keys file replaces the internal dev keys file as the internal dev would no longer be authorised. This would leave you with servers that only the external devs specified can access which would be very bad indeed.

This PR improves this role in the following ways:
- Creates an external dev children group in the `ini` file to specify which servers you want your external devs to access
- Removes the external dev keys overwrite error
- Specifies that all keys will only be appended rather than overwritten
- Uses a separate task to add external dev keys with a conditional for the task to only run against servers in the external dev `ini` group
- Uses a separate task to remove unauthorised users
- Extracts different groups of user public keys as default variables for clarity and to make it easier to use group vars to appropriately override them as necessary per server group

### CODE REVIEW

**Comment by Ocre42** [comment]
So I've had a look at this, I've updated the keys and servers list and made a couple of tweaks and there's still some issues. After some additional digging I set the user as root by default, there's still a lingering issue:

> fatal: [46.101.11.244]: FAILED! => {"msg": "The conditional check 'external_dev_servers' failed. The error was: error while evaluating conditional (external_dev_servers): 'external_dev_servers' is undefined\n\nThe error appears to have been in '/home/johna/Desktop/Projects/Third-Party/deployment/roles/configure_ssh_keys/tasks/main.yml': line 11, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Set authorized external keys\n  ^ here\n"}

It seems that `tasks/main.yml` is not picking up the variable defined in `configure_ssh_keys.yml`.

Also, if you could tick “allow edits from maintainers" I'd be able to commit my changes. 

(Also also! if you can, change it into the dev branch instead of master)

**Comment by allthatilk** [comment]
@Ocre42 Okay I'll take a look at these on my lunch break =)

**Comment by Ocre42** [comment]
Made a few changes to make it run, it doesn't seem to get the user from the items, and I made it so that unauthorised keys doesn't run unless there's at least one item in there.

File: roles/configure_ssh_keys/defaults/main.yml [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 authorized_keys_external:
   - name: shi
     public_key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCiD5YKdszmWOFDHco51vmpDOy+VjXaMQKIwSi7TCCCTFvy5dwqM0DwbHZZEgFHohT1jDqlm2adHeiHn27yv+VKfP9sGFWM7/tPGKpCTt01S/IDwmsFS/ZhgK6KI9W/YntS3QbuHkCso2rNOLDq2fxvlpOXXOMw/qDULGIKTOf+SMxmvjX4KgCQ271dE7L+OGR2wXB8l4G1bzaSFLTAXnxc0ulJATXYYOAAsY2qWhTax5tNG0K2zItiOZWffwmVKw4QwLutTCKW2qqJb8cKgawRWL48FR8cV04A8eShXF/7k9pJfKNMCMRtA1zeRGGC/zTXgyelRrSpzduy/y4ZpBEl notshi@gmail.com
   - name: kris
     public_key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDdx+ZGNaTFfqwUzphFSP7m03bu0kLYDuvVg/BXqtimIEPfYTv6K4mZPz7gpsV9OChOX70Y0KHbWVegr6pNUN0h6YxCcOlCRq5/ylifDM08K3SCd/R13MJVvbDuXiBzGflEW2WBPy1zyMEI2SR9lv5PfsRquZnjfDlejn/sLzaZ/PUVHpuweOfGdpKseCM3kh7XhwDAPwGxBx0ckafeiPZjtK0T+yiKR3SXysa2FB5UzW1WBKBqupZXrjgeQ8RC1MKK3MQAPyfiYWT57wsQqyOyPA0oVRKsoez+I3uRWp4KZm9DRya5v231Hw8YuCPWMdPrPUjkafP+RImvBHmEdC1T kriss@xixs.com
 
-unauthorised_keys: []
+unauthorised_keys: []
```
Comment (allthatilk) [inline roles/configure_ssh_keys/defaults/main.yml]: You should set your text editor up to handle stuff like this automatically. There should be a new line at the end of this file.

File: roles/configure_ssh_keys/tasks/main.yml [hunk 2]
```diff
@@ -1,28 +1,28 @@
...
   authorized_key:
-    user: "{{ item.user }}"
+    user: root
     key: "{{ item.public_key }}"
     state: absent
   with_items:
-    - unauthorised_keys
-
+    - "{{ unauthorised_keys }}"
+  when: unauthorised_keys.0 is defined
```
Comment (allthatilk) [inline roles/configure_ssh_keys/tasks/main.yml]: needs new line

File: servers.ini [hunk 3]
```diff
@@ -43,6 +43,7 @@ webserver11
 webserver14
 discourse
 
+
```
Comment (allthatilk) [inline servers.ini]: this blank line isn't needed

**Review by allthatilk — COMMENTED** [review]
Noticed some minor whitespace issues

**Comment by allthatilk** [comment]
Looks good!

**Review by Ocre42 — APPROVED** [review]
:+1:

**Pull request merged**